### PR TITLE
Vecsim search simple query

### PR DESCRIPTION
**Pull request by GuyAv46**
In this PR I updated the vector similarity top-k query flow, from after the query was parsed to the result return to the client.
I added a result processor for when the query contains `top-k` query, that loads the top-k result scores to a new score field for later sorting by.
the default score field name for a vector field "vec" is "vec_score", and it can be changed using "AS" in the query.

### CODE REVIEW

File: src/vector_index.c [hunk 1]
```diff
@@ -68,7 +68,33 @@ VecSimIndex *OpenVectorIndex(RedisSearchCtx *ctx,
   return openVectorKeysDict(ctx, keyName, 1);
 }
 
-IndexIterator *NewVectorIterator(RedisSearchCtx *ctx, VectorQuery *vq) {
+static bool isFit (VecSimIndex *ind, size_t size) {
```
Comment (DvirDukhan) [inline src/vector_index.c:71]: can we make it is a library service on a different PR?

File: src/vector_index.c [hunk 2]
```diff
@@ -95,27 +131,27 @@ int VectorQuery_EvalParams(dict *params, QueryNode *node, QueryError *status) {
...
     if (res < 0) {
       return REDISMODULE_ERR;
     }
   }
   return REDISMODULE_OK;
 }
 
-int VectorQuery_ParamResolve(VectorQueryParam *param, dict *params, QueryError *status) {
-  if (!param->isParam) {
+int VectorQuery_ParamResolve(VectorQueryParams params, size_t ix, dict *paramsDict, QueryError *status) {
```
Comment (DvirDukhan) [inline src/vector_index.c:143]: why ix?

File: src/vector_index.h [hunk 3]
```diff
@@ -32,24 +32,18 @@ typedef enum {
...
 // It is the VecSim library job to resolve this strings-key-value params (array) into a VecSimQueryParams struct.
 typedef struct {
-  const char *name;
-  size_t namelen;
-  const char *value;
-  size_t vallen;
-  bool isParam;
-} VectorQueryParam;
+  VecSimRawParam *params;
+  bool *isAttr;
```
Comment (DvirDukhan) [inline src/vector_index.h:45]: can we rename it to `shouldResolve` or something similar?

File: tests/pytests/test_search_params.py [hunk 4]
```diff
@@ -325,9 +325,9 @@ def test_vector(env):
     env.assertEqual(res2[1:], res1)
     res2 = conn.execute_command('FT.SEARCH', 'idx', '*=>[TOP_K $k @v $vec]', 'PARAMS', '4', 'vec', 'aaaaaaaa', 'k', '2', *args) 
     env.assertEqual(res2[1:], res1)
-    res2 = conn.execute_command('FT.SEARCH', 'idx', '*=>[TOP_K 2 @v $vec AS score]', 'PARAMS', '4', 'vec', 'aaaaaaaa', 'k', '2', *args) 
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '*=>[TOP_K 2 @v $vec AS v_score]', 'PARAMS', '4', 'vec', 'aaaaaaaa', 'k', '2', *args)
```
Comment (DvirDukhan) [inline tests/pytests/test_search_params.py:328]: note that this test is flawed
`aaaaaaaa` is the vector with the lowest ID. it will always return first

**Review by DvirDukhan — COMMENTED** [review]
very nice!!
few comments/questions.

File: src/result_processor.c [hunk 5]
```diff
@@ -400,10 +443,6 @@ static int rpsortNext_innerLoop(ResultProcessor *rp, SearchResult *r) {
           return RS_RESULT_ERROR;
         }
       }
-    } else if (self->sortbyType == SORTBY_DISTANCE){
```
Comment (ashtul) [inline src/result_processor.c:403]: `sortbyType` enum can be removed.
Remove also from signature of function `RPSorter_NewByFields`. Look [here](https://github.com/RediSearch/RediSearch/pull/2401/files) for change. 

It was added for the old implementation.

**Review by ashtul — COMMENTED** [review]
Nice job.
Let's try to support the special case of `*` when we parse the QueryNode tree and not on the parser. The current implementation will fail if `* =>[...]` or `(*)=>[...]`.

File: src/query.c [hunk 6]
```diff
@@ -996,6 +1004,8 @@ IndexIterator *QAST_Iterate(const QueryAST *qast, const RSSearchOptions *opts, R
 void QAST_Destroy(QueryAST *q) {
   QueryNode_Free(q->root);
   q->root = NULL;
+  array_free(q->vecScoreFieldNames);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query.c:1007]: each vector node own its score field

File: src/aggregate/aggregate_request.c [hunk 7]
```diff
@@ -928,15 +928,35 @@ static ResultProcessor *getGroupRP(AREQ *req, PLN_GroupStep *gstp, ResultProcess
   return pushRP(req, groupRP, rpUpstream);
 }
 
-#define _SCORE_LEN 6
+static ResultProcessor *getVecSimRP(AREQ *req, AGGPlan *pln, ResultProcessor *up, QueryError *status) {
+  RLookup *lk = AGPLN_GetLookup(pln, NULL, AGPLN_GETLOOKUP_LAST);
+  char **sfns = req->ast.vecScoreFieldNames;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/aggregate_request.c:933]: short for Score Field NameS. this is just a shorter name for the function. I can use `req->ast.vecScoreFieldNames` instead

File: src/aggregate/aggregate_request.c [hunk 8]
```diff
@@ -1146,6 +1145,15 @@ int AREQ_BuildPipeline(AREQ *req, int options, QueryError *status) {
   AGGPlan *pln = &req->ap;
   ResultProcessor *rp = NULL, *rpUpstream = req->qiter.endProc;
 
+  // Load vecsim results according to their score fields.
+  // We need this RP only if vecScoreFieldNames is not empty.
+  if (req->ast.vecScoreFieldNames) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/aggregate/aggregate_request.c:1150]: yes. each VN has this string so the length of this array is the number of VN in the AST.

File: src/query_parser/parser.y [hunk 9]
```diff
@@ -670,38 +675,58 @@ geo_filter(A) ::= LSQB param_any(B) param_any(C) param_any(D) param_any(E) RSQB.
...
+//   switch (B->vn.vq->type) {
+//     case VECSIM_QT_TOPK:
+//       B->vn.vq->topk.order = BY_ID;
+//       break;
+//   }
+//   A = B;
+// }
+
+query ::= STAR ARROW LSQB vector_query(B) RSQB . { // main parse, simple vecsim search as entire query case.
+  setup_trace(ctx);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query_parser/parser.y:688]: yea I just copied it from another `query` rule

File: src/query_parser/parser.y [hunk 10]
```diff
@@ -711,27 +736,33 @@ vector_command(A) ::= TOP_K param_size(B) modifier(C) ATTRIBUTE(D). {
   D.type = QT_PARAM_VEC;
   A = NewVectorNode_WithParams(ctx, VECSIM_QT_TOPK, &B, &D);
   A->vn.vq->property = rm_strndup(C.s, C.len);
+  A->vn.vq->scoreField = rm_calloc(1, C.len + 7);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query_parser/parser.y:739]: I can put it in `tokenizer.h`

File: src/vector_index.c (hunk 1 above)
Comment (GuyAv46) [reply src/vector_index.c:71]: I think we should add a check for that in the library. Maybe when we add support for more types?

File: src/vector_index.c (hunk 2 above)
Comment (GuyAv46) [reply src/vector_index.c:143]: why i :) ? short for index but I can change that

File: src/vector_index.h (hunk 3 above)
Comment (GuyAv46) [reply src/vector_index.h:45]: OK

File: src/query_parser/parser.y [hunk 11]
```diff
@@ -178,12 +178,17 @@ void reportSyntaxError(QueryError *status, QueryToken* tok, const char *msg) {
 %type vector_command { QueryNode *}
 %destructor vector_command { QueryNode_Free($$); }
 
-%type vector_attribute { VectorQueryParam }
-%destructor vector_attribute { rm_free((char*)$$.value); }
+%type vector_attribute { SingleVectorQueryParam }
+%destructor vector_attribute { rm_free((char*)($$.param.value)); rm_free((char*)($$.param.name)); }
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query_parser/parser.y:182]: OH :O
I thought the hole idea of declaring a destructor is that you don't have to free it yourself

File: src/query_parser/parser.y [hunk 12]
```diff
@@ -670,38 +675,58 @@ geo_filter(A) ::= LSQB param_any(B) param_any(C) param_any(D) param_any(E) RSQB.
...
   switch (B->vn.vq->type) {
     case VECSIM_QT_TOPK:
-      B->vn.vq->topk.runType = VECSIM_RUN_KNN;
-      B->vn.vq->topk.order = BY_ID;
+      B->vn.vq->topk.order = BY_SCORE;
       break;
   }
-  A = B;
+  ctx->root = B;
 }
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query_parser/parser.y:695]: this rule guarantees that the entire query part in the FT.SEARCH command is a vecsim query. do you mean having a filter outside the query part?

File: src/result_processor.c (hunk 5 above)
Comment (GuyAv46) [reply src/result_processor.c:403]: I thought we need it for sortby_score, Ill remove it

File: src/query.h [hunk 13]
```diff
@@ -35,6 +35,9 @@ typedef struct QueryAST {
   const void *udata;
   size_t udatalen;
 
+  // vector score field names array in the AST.
+  char **vecScoreFieldNames;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/query.h:39]: OK. But it is a property of the AST don't you think?

**Review by DvirDukhan — APPROVED** [review]

File: src/query_parser/parser.y [hunk 14]
```diff
@@ -743,9 +743,7 @@ vector_command(A) ::= TOP_K param_size(B) modifier(C) ATTRIBUTE(D). {
   D.type = QT_PARAM_VEC;
   A = NewVectorNode_WithParams(ctx, VECSIM_QT_TOPK, &B, &D);
   A->vn.vq->property = rm_strndup(C.s, C.len);
-  A->vn.vq->scoreField = rm_calloc(1, C.len + strlen(VECSIM_DEFAULT_SCORE_FIELD_SUFFIX));
-  strncpy(A->vn.vq->scoreField, C.s, C.len);
-  strcat(A->vn.vq->scoreField, VECSIM_DEFAULT_SCORE_FIELD_SUFFIX);
+  assert(-1 != (asprintf(&A->vn.vq->scoreField, "__%.*s_score", C.len, C.s)));
```
Comment (DvirDukhan) [inline src/query_parser/parser.y:746]: RedisModule_Assert
also, if malloc is used here, how do you free this string?

Comment (GuyAv46) [reply src/query_parser/parser.y:746]: it is freed in `QueryVectorNode_Free`

**Pull request merged**